pkg/providers/cloudstack/validator.go

```diff
@@ -0,0 +1,197 @@
+	}
+
+	if err = v.validateMachineConfig(ctx, clusterSpec.datacenterConfig.Spec, controlPlaneMachineConfig); err != nil {
+		logger.V(1).Info("Control plane machine config validation failed.")
+		return err
+	}
+	if controlPlaneMachineConfig.Spec.Template != workerNodeGroupMachineConfig.Spec.Template {
+		if workerNodeGroupMachineConfig.Spec.Template.Value == "" {
+			return fmt.Errorf("worker CloudStackMachineConfig template is not set. Default template is not supported in CloudStack, please provide a template name")
+		}
+		if err = v.validateMachineConfig(ctx, clusterSpec.datacenterConfig.Spec, workerNodeGroupMachineConfig); err != nil {
+			logger.V(1).Info("Workload machine config validation failed.")
+			return err
+		}
+		if controlPlaneMachineConfig.Spec.Template != workerNodeGroupMachineConfig.Spec.Template {
+			return errors.New("control plane and worker nodes must have the same template specified")
+		}
+	}
+	logger.MarkPass("Control plane and Workload templates validated")
+
+	if etcdMachineConfig != nil {
+		if etcdMachineConfig.Spec.Template.Value == "" {
+			return fmt.Errorf("etcd CloudStackMachineConfig template is not set. Default template is not supported in CloudStack, please provide a template name")
+		}
+		if err = v.validateMachineConfig(ctx, clusterSpec.datacenterConfig.Spec, etcdMachineConfig); err != nil {
+			logger.V(1).Info("Etcd machine config validation failed.")
+			return err
+		}
+		if etcdMachineConfig.Spec.Template != controlPlaneMachineConfig.Spec.Template {
+			return errors.New("control plane and etcd machines must have the same template specified")
+		}
+	}
+
+	return nil
+}
+
+func (v *Validator) validateMachineConfig(ctx context.Context, datacenterConfigSpec anywherev1.CloudStackDatacenterConfigSpec, machineConfig *anywherev1.CloudStackMachineConfig) error {
+	domain := datacenterConfigSpec.Domain
+	zone := datacenterConfigSpec.Zone
+	account := datacenterConfigSpec.Account
+	if err := v.cmk.ValidateTemplatePresent(ctx, domain, zone, account, machineConfig.Spec.Template); err != nil {
+		return fmt.Errorf("error validating template: %v", err)
+	}
+	if err := v.cmk.ValidateServiceOfferingPresent(ctx, domain, zone, account, machineConfig.Spec.ComputeOffering); err != nil {
+		return fmt.Errorf("error validating template: %v", err)
+	}
+	if len(machineConfig.Spec.AffinityGroupIds) > 0 {
+		if err := v.cmk.ValidateAffinityGroupsPresent(ctx, domain, zone, account, machineConfig.Spec.AffinityGroupIds); err != nil {
+			return fmt.Errorf("error validating affinity group ids: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func (v *Validator) validateControlPlaneHost(pHost *string) error {
+	_, port, err := net.SplitHostPort(*pHost)
+	if err != nil {
+		if strings.Contains(err.Error(), "missing port") {
+			port = controlEndpointDefaultPort
+			*pHost = fmt.Sprintf("%s:%s", *pHost, port)
```

Comment:
I am not sure how I feel about editing a value from the parameter, but I think we do something like this with ssh auth keys in vsphere. Maybe we can just call this `validateAndParseControlPlaneHost` so that it is clear that the value is being modified.

Reply:
I think you're on the right track, Vivek. This method shouldn't really modify the input, even if we do it elsewhere. Its more idiomatic to return the value and set it explicitly in the calling function. The function could do with reflecting the fact it will apply a default and a short code comment would be useful to clarify it: `validateControlPlaneHostAndApplyDefaultPort(host string) (string, error)`.

Reply:
I would advise to keep defaults outside of the validator
For vsphere we had a "validator" and a "defaulter" (no need to follow the same)
A couple of reasons:
* Changing the input in a validator is an unexpected side effect. This could lead to hidden bugs
* When running this in a validation webhook, we can't really modify the api object

Reply:
I agree - would it be sufficient to add a TODO to move this to the cloudstack provider, after this code is merged?